doc/i2c_bus_lock_prevention/i2c_bus_lock_prevention_hld.md

```diff
@@ -0,0 +1,425 @@
+
+#### 7.2. The Method to Detect I2C Bus Lock
+
+On a specific hardware model, choose three representative i2c devices to poll.
+
+The criteria to choose the representative i2c devices are shown below.
+* In order to minimize the impact to the system performance, the chosen i2c device is seldom visited by the platform driver daemons.
+* The polling interval shall be defined to a value that will not impact the system performance severely.
+
+When i2c bus lock happens, the error realized by the i2c driver is timeout error. There are two possible causes listed below for the timeout error.
+    1. The i2c slave device is dead so it does not response to the i2c command issued from the i2c controller(i.e. master device)
+    2. The i2c bus is locked. So the i2c slave device cannot receive the i2c command issued from the i2c controller.
+
+In order to detect the condition 2 correctly, it is required to poll three i2c devices to confirm that the i2c bus is REALLY locked. Given that only one i2c device is polled and it happens the device is broken and not respond to the i2c controller(i.e. condition 1 listed above). It gets the false alarm because only the i2c slave device chosen to poll is broken and not response. Actually, the i2c bus is still good and the i2c controller is able to issue i2c transactions to other i2c devices on the bus.
+
+The i2c bus is determined to be in lock state when all of the chosen i2c devices cannot be accessed correctly.
+
+#### 7.3. The Method to Recover I2C Bus Lock State
+
+When the i2c bus lock happens, the CPU triggers the reset signal to i2c mux devices to close all of the channels on these mux devices.
+
+#### 7.4. General Pseudo Design
+
+* Monitoring i2c bus health status
+    - Log when i2c bus health status is changed to bad or changed to good.
+
+* Handling i2c bus lock
+    - Reset I2C muxe devices to recover the lock state.
+    - It is known that the thermal sensors readings are accessed through the i2c bus. The fan speed control is also done through i2c bus. When i2c bus is locked, the daemon to enforce thermal policy is not working anymore. To alleviate the risk of the hazard of overheating, this design invokes the special script to set all of the fans on the switch device to full speed as long as the i2c bus health is back to normal. Given that the i2c operations to set fan speed will not be affected by the faulty i2c device, all of the fan speed can be set as full speed running.
+    - Perform the algorithm to find out the faulty device and put into the isolation list. Note that the algorithm skips checking on the device that is known to be faulty. In other words, the algorithm only scans on the healthy devices according to the existing isolation list. Only on the first occurrence of the i2c bus lock, all of the i2c devices will be scanned.
+    - After the isolation list is generated and the i2c bus health is back to normal. Recover the execution of the daemons that need to access i2c bus.

```

Comment:
How to determine which daemons need to be recovered after the device(s) are reset? Should the daemons register that they are using a particular device? That would allow for finer-grained control of which services get affected by a lockup.
If we just blindly restart `pmon` docker, that would be very disruptive and it would disrupt traffic.